remove Jetpack dependencies from the Connection package

leogermani opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

This PR removes some Jetpack-specific XMLRPC endpoints from the Connection package and places it in the Jetpack plugin.

This should not change anything for sites with the Jetpack plugin active.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Removes Jetpack specific XMLRPC endpoints from the Connection package

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
1164484600834010-as-1199928370054538

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Checkout this branch
* Connect the site
* Run the Jetpack Debugger and make sure Everything looks great
* In the debugger, scroll down to "Tests"
* Click on "success: test authenticated xmlrpc request for jetpack passes connection test"
* Make sure the response is the Jetpack version, and not the `success` string
* Test requests to the json_api endpoint
* Go to Jetpack Dashboard and Disconnect. Make sure disconnection still works


#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
<!-- Guidelines: https://github.com/Automattic/jetpack/blob/master/docs/writing-a-good-changelog-entry.md -->
* Jetpack plugin extends the Connection package and registers its own XMLRPC methods

leogermani commented on projects/plugins/jetpack/class-jetpack-xmlrpc-methods.php:
```diff
@@ -0,0 +1,96 @@
+	 */
+	public function disconnect_blog() {
+
+		/**
```
We don't need that `wp_set_current_user()` bit here because the user is already set at `Jetpack_XMLRPC_Server::login` -> https://github.com/Automattic/jetpack/blob/update%2Fconnection-package-xmlrpc-independence/projects/packages/connection/legacy/class-jetpack-xmlrpc-server.php#L600

sergeymitr reviewed (approved):
Works well 👍

jeherve reviewed (dismissed):
It looks like the tests are failing at the moment because of the removed methods.

leogermani commented:
Thanks for catching this @jeherve 

This small failing test took me into a small rabbit hole.

1. Since the test was calling the class method directly, I guess it's better to keep the methods there and deprecate them, just in case. I've seen cases where other plugin or themes use things that they shouldn't.

So https://github.com/Automattic/jetpack/pull/18952/commits/2b60c752a9e31e4413748085ac49a8efc7989bd3 adds the deprecate methods

2. While looking at the tests, I've found another method that should be moved to the plugin

https://github.com/Automattic/jetpack/pull/18952/commits/24b7b43ae3438aa82e85eda23dc2fac4aaf4142d moves `remote_provision` to the plugin.

3. Then I went to see these tests and, since they are testing the class that lives in the Connection package, they should also live in the package.

So I moved them to the package while leaving some jetpack specific tests in the plugin.

https://github.com/Automattic/jetpack/pull/18952/commits/1dbda8beb79ec44517a3fdc05899ed8ba068279f does that

4. Then I also

* Fixed a test calling a deprecated function
* https://github.com/Automattic/jetpack/pull/18952/commits/45edc5ae9cafc1302f3fe908cce96c39bc4b064e
* added changelog https://github.com/Automattic/jetpack/pull/18952/commits/d9b1fa710c2fc9d0496438e9c2884241f29d2bcf
* juggled with phpunit version so the tests pass: https://github.com/Automattic/jetpack/pull/18952/commits/40d7105bf72afb9eb01fc2f12d54065fbd039c74 and https://github.com/Automattic/jetpack/pull/18952/commits/da73201bf575619dced37542a9675040c7a7f749

Now wait for the jobs to complete to make sure it's working

leogermani commented:
@jeherve @sergeymitr PR is good for review again

anomiex commented on projects/packages/connection/tests/php/test_jetpack_xmlrpc_server.php:
```diff
@@ -0,0 +1,438 @@
+		);
+		( new Tokens() )->update_user_token( 1, sprintf( '%s.%d', 'token', 1 ), true );
+
+		$server = new Jetpack_XMLRPC_Server();
+
+		$response = $server->remote_connect(
+			array(
+				'nonce'      => '1234',
+				'local_user' => $this->xmlrpc_admin,
+			)
+		);
+
+		$this->assertInstanceOf( 'IXR_Error', $response );
+		$this->assertObjectHasAttribute( 'code', $response );
+		$this->assertObjectHasAttribute( 'message', $response );
+		$this->assertEquals( 400, $response->code );
+		$this->assertEquals(
+			'Jetpack: [token_fetch_failed] Failed to fetch user token from WordPress.com.',
+			$response->message
+		);
+
+		foreach ( array( 'blog_token', 'id', 'master_user', 'user_tokens' ) as $option_name ) {
+			Jetpack_Options::delete_option( $option_name );
+		}
```
I note this cleanup won't happen if any of the tests fail. If the cleanup is actually needed, you should probably do something like this to make sure it happens.
```suggestion
		try {
			$server = new Jetpack_XMLRPC_Server();

			$response = $server->remote_connect(
				array(
					'nonce'      => '1234',
					'local_user' => $this->xmlrpc_admin,
				)
			);

			$this->assertInstanceOf( 'IXR_Error', $response );
			$this->assertObjectHasAttribute( 'code', $response );
			$this->assertObjectHasAttribute( 'message', $response );
			$this->assertEquals( 400, $response->code );
			$this->assertEquals(
				'Jetpack: [token_fetch_failed] Failed to fetch user token from WordPress.com.',
				$response->message
			);
		} finally {
			foreach ( array( 'blog_token', 'id', 'master_user', 'user_tokens' ) as $option_name ) {
				Jetpack_Options::delete_option( $option_name );
			}
		}
```

anomiex commented on projects/packages/connection/tests/php/test_jetpack_xmlrpc_server.php:
```diff
@@ -0,0 +1,438 @@
+			->method( 'isError' )
+			->will( $this->returnValue( empty( $error ) ? false : true ) );
+
+		$xml->expects( $this->exactly( empty( $error ) ? 1 : 0 ) )
```
You can feel free to ignore, but unnecessary use of `empty()` is one of my pet peeves. p1614091140044600-slack-C6JKZ8MLK
```suggestion
		$xml->expects( $this->exactly( $error ? 0 : 1 ) )
```

anomiex reviewed (commented):
Gave it a once-over. I'd recommend at least fixing the comment on line 147; the rest I don't care so much about.

leogermani commented on projects/packages/connection/tests/php/test_jetpack_xmlrpc_server.php:
```diff
@@ -0,0 +1,438 @@
+			$this->assertInstanceOf( 'IXR_Error', $response );
+			$this->assertEquals( $expected['code'], $response->code );
+
+			// Not using assertContains or assertStringContainsString to work with all phpunit versions.
```
Thanks @anomiex , I'll address your suggestions.

For the record, with the exception of the line of code in this comment, all the rest is very old code I just moved around. :)

leogermani replied:
done

anomiex commented on projects/packages/connection/legacy/class-jetpack-xmlrpc-server.php:
```diff
@@ -98,14 +89,15 @@ public function xmlrpc_methods( $core_methods ) {
-		 * @param array $core_methods    Available core XML-RPC methods.
+		 * @param array         $jetpack_methods XML-RPC methods available to the Jetpack Server.
+		 * @param array         $core_methods    Available core XML-RPC methods.
+		 * @param \WP_User|bool $user            Information about a given WordPress user. False if authenticated with blog token.
```
Do we document in which version a parameter was added to a filter?

leogermani replied:
Yes, thanks for the reminder. Done in https://github.com/Automattic/jetpack/pull/18952/commits/21418d1e5f212492038290abfc17c8a7989ea123

anomiex reviewed (approved):
I node that `jetpack.remoteRegister` and `jetpack.remoteConnect` still call methods on the Jetpack class. Maybe that's out of scope for this PR though.

Haven't manually tested, but it looks like there is test coverage.

leogermani commented:
> I node that `jetpack.remoteRegister` and `jetpack.remoteConnect` still call methods on the Jetpack class. Maybe that's out of scope for this PR though.
> 
> Haven't manually tested, but it looks like there is test coverage.

I've checked all the `remote_*` methods and asked for confirmation. (see p1615909186005600-slack-CDLH4C1UZ). As I suspected, these are features that the Connection package must provide, so it's a job for another PR to remove the Jetpack dependency for those methods.

I reverted the commit in which I moved `remote_provision` method to the plugin.

leogermani commented:
@sergeymitr @jeherve could you have a look at this again? I'd love to get it merged. I'm touching some files in another PR that will likely conflict and I'd also want to use the new test files in the Conncetion package to add some tests in the other PR

sergeymitr commented:
Hi @leogermani.

Methods in the class `Jetpack_XMLRPC_Methods` are getting called as static, which they are not, so we get errors like this in the `debug.log`:
```
PHP Deprecated:  Non-static method Jetpack_XMLRPC_Methods::json_api() should not be called statically in /var/www/html/wp-includes/IXR/class-IXR-server.php on line 121
```

It's true for methods `test_connection()`, `disconnect_blog()`, `json_api()`.
If you switch them to static, please make sure that they aren't called as object methods anywhere else.

leogermani commented:
> Hi @leogermani.
> 
> Methods in the class `Jetpack_XMLRPC_Methods` are getting called as static, which they are not, so we get errors like this in the `debug.log`:
> 
> ```
> PHP Deprecated:  Non-static method Jetpack_XMLRPC_Methods::json_api() should not be called statically in /var/www/html/wp-includes/IXR/class-IXR-server.php on line 121
> ```
> 
> It's true for methods `test_connection()`, `disconnect_blog()`, `json_api()`.
> If you switch them to static, please make sure that they aren't called as object methods anywhere else.

Thanks for spotting this @sergeymitr . They should be static, yes. I fixed it.

I also removed an old debugging comment that was there.

sergeymitr reviewed (approved):
👍
